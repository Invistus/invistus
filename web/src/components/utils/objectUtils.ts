export const isNotEmpty = (obj: any): boolean => {
    return obj !== undefined && Object.keys(obj).length > 0;
};
