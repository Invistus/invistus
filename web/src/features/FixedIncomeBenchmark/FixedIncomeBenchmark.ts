export type InvestmentType = 'CDB' | 'LCA/LCI';
export type InvestmentCategory = 'pre_fixed' | 'inflation_linked' | 'post_fixed';

export interface IFixedIncomeBenchmark {
  investmentType: InvestmentType,
  investmentCategory: InvestmentCategory,
  investmentCategoryIndexPostFixed: number,
  investmentCategoryIndexIPCA: number,
  grossReturnPercentage: number,
  durationDays: number,
  dueDate?: Date
};

export type SourceResult = {
  investmentType: InvestmentType;
  investmentCategory: InvestmentCategory;
  grossReturnPercentage: number;
  durationDays: number;
  tax_range?: string;
  tax_rate?: number;
  netReturn: number;
};

export type BenchmarkResult = {
  investmentType: InvestmentType;
  investmentCategory: InvestmentCategory;
  investmentCategoryIndexPostFixed?: number;
  investmentCategoryIndexIPCA?: number;
  grossReturnPercentage: number;
  durationDays: number;
  tax_range?: string;
  tax_rate?: number;
  netReturn: number;
};

export type Result = {
  source: SourceResult;
  benchmark: BenchmarkResult[];
};

const roundPercentage = (percentage: number): number => {
  return Number(percentage.toFixed(4)) 
}
 
const calculateTaxRate = (days: number): [string, number] => {
  if (days <= 180) return ['0_180_days', 0.225];
  if (days <= 360) return ['181_360_days', 0.20];
  if (days <= 720) return ['361_720_days', 0.175];
  return ['above_720_days', 0.15];
};

const calculateNetReturn = (type: InvestmentType, category: InvestmentCategory, grossReturn: number, durationDays: number): number => {
  if (type === 'CDB') {
    const tax = calculateTaxRate(durationDays)[1];
    return grossReturn * (1 - tax);
  }
  return grossReturn;
};

const calculateGrossReturnForSameNetReturn = (type: InvestmentType, category: InvestmentCategory, netReturn: number, data: IFixedIncomeBenchmark): number => {
  if (type === 'CDB') {
    const tax = calculateTaxRate(data.durationDays)[1];
    netReturn = netReturn / (1 - tax);
  }

  switch (category) {
    case 'pre_fixed':
      return netReturn;
    case 'inflation_linked':
      return netReturn - data.investmentCategoryIndexIPCA;
    case 'post_fixed':
      if (data.investmentCategoryIndexPostFixed !== 0) {
        return netReturn / data.investmentCategoryIndexPostFixed;
      }
      return 0; // Handle division by zero
  }
};

const calculateSourceGrossReturn = (data: IFixedIncomeBenchmark): number => {
  switch (data.investmentCategory) {
    case 'pre_fixed':
      return data.grossReturnPercentage;
    case 'inflation_linked':
      return data.investmentCategoryIndexIPCA + data.grossReturnPercentage;
    case 'post_fixed':
      return data.investmentCategoryIndexPostFixed * data.grossReturnPercentage;
  }
}

export function validateFixedIncomeBenchmark(data: IFixedIncomeBenchmark): void {
  if (data.investmentCategoryIndexPostFixed < 0) {
      throw new Error('Investment Category Index Post Fixed cannot be negative.');
  }

  if (data.investmentCategoryIndexIPCA < 0) {
      throw new Error('Investment Category Index IPCA cannot be negative.');
  }

  if (data.grossReturnPercentage < 0) {
      throw new Error('Gross Return Percentage cannot be negative.');
  }

  if (data.durationDays < 0) {
      throw new Error('Duration Days cannot be negative.');
  }
  
  if (data.dueDate && data.dueDate < new Date()) {
      throw new Error('Due Date cannot be in the past.');
  }
}

export function calculateBenchmark(data: IFixedIncomeBenchmark): Result {

  // Validate data
  validateFixedIncomeBenchmark(data);

  // Calculate source total return
  const sourceGrossReturn = calculateSourceGrossReturn(data);
  const sourceNetReturn = calculateNetReturn(data.investmentType, data.investmentCategory, sourceGrossReturn, data.durationDays);
  const [taxRange, taxRate] = calculateTaxRate(data.durationDays);

  const sourceResult: SourceResult = {
    investmentType: data.investmentType,
    investmentCategory: data.investmentCategory,
    grossReturnPercentage: roundPercentage(data.grossReturnPercentage),
    durationDays: data.durationDays,
    netReturn: roundPercentage(sourceNetReturn)
  };

  if (data.investmentType === 'CDB') {
    sourceResult.tax_range = taxRange;
    sourceResult.tax_rate = taxRate;
  }

  const allTypes: InvestmentType[] = ['CDB', 'LCA/LCI'];
  const allCategories: InvestmentCategory[] = ['pre_fixed', 'inflation_linked', 'post_fixed'];

  // Calculate benchmark
  const benchmarks: BenchmarkResult[] = allTypes.flatMap(type => 
    allCategories.filter(category => !(type === data.investmentType && category === data.investmentCategory))
    .map(category => {
      const benchmarkGrossReturn = calculateGrossReturnForSameNetReturn(type, category, sourceNetReturn, data);

      const benchmark: BenchmarkResult = {
        investmentType: type,
        investmentCategory: category,
        grossReturnPercentage: roundPercentage(benchmarkGrossReturn),
        durationDays: data.durationDays,
        netReturn: roundPercentage(sourceNetReturn)
      };

      if (category === 'post_fixed') {
        benchmark.investmentCategoryIndexPostFixed = data.investmentCategoryIndexPostFixed;
      }

      if (category === 'inflation_linked') {
        benchmark.investmentCategoryIndexIPCA = data.investmentCategoryIndexIPCA;
      }

      if (type === 'CDB') {
        benchmark.tax_range = taxRange;
        benchmark.tax_rate = taxRate;
      }

      return benchmark;
    })
  );

  return {
    source: sourceResult,
    benchmark: benchmarks
  };
}
