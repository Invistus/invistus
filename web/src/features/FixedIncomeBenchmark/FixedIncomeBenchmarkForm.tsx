import React from 'react';
import { useForm, FieldValues, FieldErrors } from 'react-hook-form';
import { parsePercentage, parseNumber } from 'utils/stringUtils';
import PercentageField from 'components/forms/fields/PercentageField';
import NumberField from 'components/forms/fields/NumberField';
import { InvestmentType, InvestmentCategory, IFixedIncomeBenchmark } from 'features/FixedIncomeBenchmark/FixedIncomeBenchmark';
import "react-datepicker/dist/react-datepicker.css";
import DateField from 'components/forms/fields/DateField';
import RadioGroupField from 'components/forms/fields/RadioGroupField';
import ButtonGroup from 'components/forms/groups/ButtonGroup';
import Submit from 'components/forms/inputs/Submit';
import Form from 'components/forms/Form';
import { isNotEmpty } from 'utils/objectUtils';
import useTranslation from 'language/useTranslation';
import { getInputDefaults } from 'utils/formUtils';


interface FixedIncomeBenchmarkInputValues extends FieldValues {
  investmentType: InvestmentType;
  investmentCategory: InvestmentCategory;
  investmentCategoryIndexPostFixed: string,
  investmentCategoryIndexIPCA: string,
  grossReturnPercentage: string;
  durationDays: string;
  dueDate?: Date;
};

const { number, percentage } = getInputDefaults();
const defaultFixedIncomeBenchmarkInputValues: FixedIncomeBenchmarkInputValues =  {
  investmentType: 'CDB',
  investmentCategory: 'pre_fixed',
  investmentCategoryIndexPostFixed: percentage,
  investmentCategoryIndexIPCA: percentage,
  durationDays: number,
  grossReturnPercentage: percentage,
  dueDate: new Date()
}

const FixedIncomeBenchmarkForm: React.FC<{ onSubmit: (data: IFixedIncomeBenchmark) => void }> = ({ onSubmit }) => {

  const { t } = useTranslation();

  const { control, handleSubmit, setValue, formState: { errors } } = useForm<FixedIncomeBenchmarkInputValues, any, IFixedIncomeBenchmark>({
    defaultValues: defaultFixedIncomeBenchmarkInputValues,
    resolver: async (inputValues: FixedIncomeBenchmarkInputValues) => {

      const formData: IFixedIncomeBenchmark = {
        investmentType: inputValues.investmentType,
        investmentCategory: inputValues.investmentCategory,
        investmentCategoryIndexPostFixed: parsePercentage(inputValues.investmentCategoryIndexPostFixed) || 0,
        investmentCategoryIndexIPCA: parsePercentage(inputValues.investmentCategoryIndexIPCA) || 0,
        durationDays: parseNumber(inputValues.durationDays) || 0,
        grossReturnPercentage: parsePercentage(inputValues.grossReturnPercentage) || 0,
        dueDate: inputValues.dueDate,
      };

      const errors: FieldErrors<FixedIncomeBenchmarkInputValues> = {};
      
      if (formData.investmentCategoryIndexPostFixed <= 0) {
        errors.investmentCategoryIndexPostFixed = {
            type: "manual",
            message: t('fixedIncomeBenchmark.errors.investmentCategoryIndexPostFixedGreaterThanZero')
          };    
      }   

      if (formData.investmentCategoryIndexIPCA <= 0) {
        errors.investmentCategoryIndexIPCA = {
            type: "manual",
            message: t('fixedIncomeBenchmark.errors.investmentCategoryIndexIPCAGreaterThanZero')
          };    
      }  

      if (formData.grossReturnPercentage <= 0) {
        errors.grossReturnPercentage = {
            type: "manual",
            message: t('fixedIncomeBenchmark.errors.grossReturnPercentageGreaterThanZero')
          };    
      }  

      if (formData.durationDays <= 0) {
        errors.durationDays = {
            type: "manual",
            message: t('fixedIncomeBenchmark.errors.durationDaysGreaterThanZero')
          };    
      }  

      return {
        values: isNotEmpty(errors) ? {} : formData,
        errors: errors,
      };
    },
  });
  

  const onChangeDueDate = (dueDate: Date | null) => {
    const durationDays = Math.ceil(((dueDate ?? new Date()).getTime() - new Date().getTime()) / (1000 * 60 * 60 * 24));
    setValue('durationDays', (durationDays > 0 ? durationDays : 1).toString());
  }

  const onChangeDurationDays = (durationDays: number) => {
    const newDate = new Date();
    newDate.setDate(newDate.getDate() + Number(durationDays));
    setValue('dueDate', newDate);
  }

  return (
    <Form onSubmit={handleSubmit(onSubmit)} errors={errors}>
     
      <RadioGroupField
          label={t('fixedIncomeBenchmark.investmentType')}
          name="investmentType"
          control={control}
          options={[
            { label: t('fixedIncomeBenchmark.cdb'), value: 'CDB' },
            { label: t('fixedIncomeBenchmark.lca_lci'), value: 'LCA/LCI' }
          ]}>
      </RadioGroupField>

      <RadioGroupField
          label={t('fixedIncomeBenchmark.investmentCategory')}
          name="investmentCategory"
          control={control}
          options={[
            { label: t('fixedIncomeBenchmark.pre_fixed'), value: "pre_fixed" },
            { label: t('fixedIncomeBenchmark.post_fixed'), value: "post_fixed" },
            { label: t('fixedIncomeBenchmark.inflation_linked'), value: "inflation_linked" }
          ]}>
      </RadioGroupField>

      <PercentageField
            label={t('fixedIncomeBenchmark.investmentCategoryIndexPostFixed')}
            name="investmentCategoryIndexPostFixed"
            control={control}
          />

        <PercentageField
            label={t('fixedIncomeBenchmark.investmentCategoryIndexIPCA')}
            name="investmentCategoryIndexIPCA"
            control={control}
          />

      <PercentageField
            label={t('fixedIncomeBenchmark.grossReturnPercentage')}
            name="grossReturnPercentage"
            control={control}
          />

      <NumberField
            label={t('fixedIncomeBenchmark.durationDays')}
            name="durationDays"
            onChange={onChangeDurationDays}
            control={control}
          />

      <DateField
            label={t('fixedIncomeBenchmark.dueDate')}
            name="dueDate"
            inputId="dueDate"
            minDate={new Date()}
            control={control}
            onChange={onChangeDueDate}    
          />

      <ButtonGroup>
        <Submit label={t('fixedIncomeBenchmark.submit')} />
      </ButtonGroup>

    </Form>
  );
};

export default FixedIncomeBenchmarkForm;
