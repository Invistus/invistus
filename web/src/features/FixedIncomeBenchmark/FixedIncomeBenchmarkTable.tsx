import React from 'react';
import { formatPercentage } from 'utils/stringUtils';
import useTranslation from 'language/useTranslation';
import { Result } from './FixedIncomeBenchmark';

interface FixedIncomeBenchmarkTableProps {
  data: Result;
}

export const FixedIncomeBenchmarkTable: React.FC<FixedIncomeBenchmarkTableProps> = ({ data }) => {

  const { t } = useTranslation();

  return (
      <div className="container table">
        <div className="row">
            <div className="col-2">{t('fixedIncomeBenchmark.investmentType')}</div>
            <div className="col-2">{t('fixedIncomeBenchmark.investmentCategory')}</div>
            <div className="col-2">{t('fixedIncomeBenchmark.dividendYield')}</div>
            <div className="col-2">{t('fixedIncomeBenchmark.tax_table')}</div>
            <div className="col-2">{t('fixedIncomeBenchmark.tax')}</div>
            <div className="col-2">{t('fixedIncomeBenchmark.netReturn')}</div>
          </div>
          <div className="row">
            <div className="col-2">{data.source.investmentType}</div>
            <div className="col-2">{t(`fixedIncomeBenchmark.${data.source.investmentCategory}`)}</div>
            <div className="col-2 number">{formatPercentage(data.source.grossReturnPercentage)}</div>
            <div className="col-2 center">{data.source.tax_range && t(`fixedIncomeBenchmark.tax_${data.source.tax_range}`)}</div>
            <div className="col-2 number">{data.source.tax_rate && formatPercentage(data.source.tax_rate)}</div>
            <div className="col-2 number">{formatPercentage(data.source.netReturn)}</div>
          </div>
          {data.benchmark.map((benchmark, index) => 
            <div className="row" key={index}>
              <div className="col-2">{benchmark.investmentType}</div>
              <div className="col-2">{t(`fixedIncomeBenchmark.${benchmark.investmentCategory}`)}</div>
              <div className="col-2 number">{formatPercentage(benchmark.grossReturnPercentage)}</div>
              <div className="col-2 center">{benchmark.tax_range ? t(`fixedIncomeBenchmark.tax_${benchmark.tax_range}`) : '-'}</div>
              <div className="col-2 number">{benchmark.tax_rate ? formatPercentage(benchmark.tax_rate) : '-'}</div>
              <div className="col-2 number">{formatPercentage(benchmark.netReturn)}</div>
            </div>                 
          )}   
    </div>
  );
};