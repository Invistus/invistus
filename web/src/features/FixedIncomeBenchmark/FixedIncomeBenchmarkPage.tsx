import React, { useState } from 'react';
import FixedIncomeBenchmarkForm from 'features/FixedIncomeBenchmark/FixedIncomeBenchmarkForm';
import { Result, calculateBenchmark } from 'features/FixedIncomeBenchmark/FixedIncomeBenchmark';
import { formatPercentage } from 'utils/stringUtils';
import { IFixedIncomeBenchmark } from 'features/FixedIncomeBenchmark/FixedIncomeBenchmark';
import Panel from 'components/panels/Panel';
import Page from 'components/pages/Page';
import useTranslation from 'language/useTranslation';
import { FixedIncomeTutorial } from '../Tutorials/FixedIncomeTutorial';
import TutorialPanel from 'features/Tutorials/TutorialPanel';


export const FixedIncomeBenchmarkPage: React.FC = () => {

  const { t } = useTranslation();
  const [result, setResult] = useState<Result>();

  const handleFormSubmit = async (data: IFixedIncomeBenchmark) => {
    const r = await calculateBenchmark(data);
    setResult(r);
  };

  return (
    <Page>
      <Panel title={t('fixedIncomeBenchmark.label')} image="/images/assets/fixed_income_benchmark.png">
        <FixedIncomeBenchmarkForm onSubmit={handleFormSubmit} />
      </Panel>
      {result && 
        <>
          <Panel>
            <h3>{t('fixedIncomeBenchmark.resultLabel').replace('[durationDays]', result.source.durationDays.toString())}</h3>
            <div className="container table">
              <div className="row">
                  <div className="col">{t('fixedIncomeBenchmark.investmentType')}</div>
                  <div className="col">{t('fixedIncomeBenchmark.investmentCategory')}</div>
                  <div className="col">{t('fixedIncomeBenchmark.dividendYield')}</div>
                  <div className="col">{t('fixedIncomeBenchmark.tax_table')}</div>
                  <div className="col">{t('fixedIncomeBenchmark.tax')}</div>
                  <div className="col">{t('fixedIncomeBenchmark.netReturn')}</div>
                </div>
                <div className="row">
                  <div className="col">{result.source.investmentType}</div>
                  <div className="col">{t(`fixedIncomeBenchmark.${result.source.investmentCategory}`)}</div>
                  <div className="col number">{formatPercentage(result.source.grossReturnPercentage)}</div>
                  <div className="col center">{result.source.tax_range && t(`fixedIncomeBenchmark.tax_${result.source.tax_range}`)}</div>
                  <div className="col number">{result.source.tax_rate && formatPercentage(result.source.tax_rate)}</div>
                  <div className="col number">{formatPercentage(result.source.netReturn)}</div>
                </div>
                {result.benchmark.map((benchmark, index) => 
                  <div className="row" key={index}>
                    <div className="col">{benchmark.investmentType}</div>
                    <div className="col">{t(`fixedIncomeBenchmark.${benchmark.investmentCategory}`)}</div>
                    <div className="col number">{formatPercentage(benchmark.grossReturnPercentage)}</div>
                    <div className="col center">{benchmark.tax_range ? t(`fixedIncomeBenchmark.tax_${benchmark.tax_range}`) : '-'}</div>
                    <div className="col number">{benchmark.tax_rate ? formatPercentage(benchmark.tax_rate) : '-'}</div>
                    <div className="col number">{formatPercentage(benchmark.netReturn)}</div>
                  </div>                 
                )}   
          </div>
        </Panel>
        <div className="more_about">{t('common.more_about')}</div>
        <TutorialPanel tutorial={<FixedIncomeTutorial />} />
      </>
      }
    </Page>
  );
};