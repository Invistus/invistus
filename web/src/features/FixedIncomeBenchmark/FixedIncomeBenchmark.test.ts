import { calculateBenchmark, IFixedIncomeBenchmark, Result } from './FixedIncomeBenchmark';  // Adjust path accordingly
import { formatPercentage } from '../../utils/stringUtils';

describe('FixedIncomeBenchmark', () => {

    describe('calculateBenchmark', () => {

        const baseData = {
            investmentCategoryIndexPostFixed: .1275,
            investmentCategoryIndexIPCA: .06,
        };

        it('should calculate source and benchmarks correctly for CDB with pre_fixed category', () => {
            const sampleData: IFixedIncomeBenchmark = {
                investmentType: 'CDB',
                investmentCategory: 'pre_fixed',
                grossReturnPercentage: .0927,
                durationDays: 731,
                ...baseData
            };

            const result: Result = calculateBenchmark(sampleData);

            const expectedSourceResult = {
                "source": {
                    "investmentType": "CDB",
                    "investmentCategory": "pre_fixed",
                    "grossReturnPercentage": 0.0927,
                    "durationDays": 731,
                    "netReturn": 0.0788,
                    "tax_range": "above_720_days",
                    "tax_rate": 0.15
                },
                "benchmark": [
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": 0.0327,
                        "durationDays": 731,
                        "netReturn": 0.0788,
                        "investmentCategoryIndexIPCA": 0.06,
                        "tax_range": "above_720_days",
                        "tax_rate": 0.15
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.7271,
                        "durationDays": 731,
                        "netReturn": 0.0788,
                        "investmentCategoryIndexPostFixed": 0.1275,
                        "tax_range": "above_720_days",
                        "tax_rate": 0.15
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.0788,
                        "durationDays": 731,
                        "netReturn": 0.0788
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": 0.0188,
                        "durationDays": 731,
                        "netReturn": 0.0788,
                        "investmentCategoryIndexIPCA": 0.06
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.618,
                        "durationDays": 731,
                        "netReturn": 0.0788,
                        "investmentCategoryIndexPostFixed": 0.1275
                    }
                ]
            };

            expect(result).toEqual(expectedSourceResult);
        });

        it('should calculate source and benchmarks correctly for CDB with inflation_linked category', () => {
            const sampleData: IFixedIncomeBenchmark = {
                investmentType: 'CDB',
                investmentCategory: 'inflation_linked',
                grossReturnPercentage: .0927,
                durationDays: 280,
                ...baseData
            };

            const result: Result = calculateBenchmark(sampleData);
            
            const expectedSourceResult = {
                "source": {
                    "investmentType": "CDB",
                    "investmentCategory": "inflation_linked",
                    "grossReturnPercentage": 0.0927,
                    "durationDays": 280,
                    "netReturn": 0.1222,
                    "tax_range": "181_360_days",
                    "tax_rate": 0.2
                },
                "benchmark": [
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.1527,
                        "durationDays": 280,
                        "netReturn": 0.1222,
                        "tax_range": "181_360_days",
                        "tax_rate": 0.2
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 1.1976,
                        "durationDays": 280,
                        "netReturn": 0.1222,
                        "investmentCategoryIndexPostFixed": 0.1275,
                        "tax_range": "181_360_days",
                        "tax_rate": 0.2
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.1222,
                        "durationDays": 280,
                        "netReturn": 0.1222
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": 0.0622,
                        "durationDays": 280,
                        "netReturn": 0.1222,
                        "investmentCategoryIndexIPCA": 0.06
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.9581,
                        "durationDays": 280,
                        "netReturn": 0.1222,
                        "investmentCategoryIndexPostFixed": 0.1275
                    }
                ]
            };

            expect(result).toEqual(expectedSourceResult);
        });

        it('should calculate source and benchmarks correctly for CDB with post_fixed category', () => {
            const sampleData: IFixedIncomeBenchmark = {
                investmentType: 'CDB',
                investmentCategory: 'post_fixed',
                grossReturnPercentage: .0927,
                durationDays: 120,
                ...baseData
            };

            const result: Result = calculateBenchmark(sampleData);
            
            const expectedSourceResult = {
                "source": {
                    "investmentType": "CDB",
                    "investmentCategory": "post_fixed",
                    "grossReturnPercentage": 0.0927,
                    "durationDays": 120,
                    "netReturn": 0.0092,
                    "tax_range": "0_180_days",
                    "tax_rate": 0.225
                },
                "benchmark": [
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.0118,
                        "durationDays": 120,
                        "netReturn": 0.0092,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": -0.0482,
                        "durationDays": 120,
                        "netReturn": 0.0092,
                        "investmentCategoryIndexIPCA": 0.06,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.0092,
                        "durationDays": 120,
                        "netReturn": 0.0092
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": -0.0508,
                        "durationDays": 120,
                        "netReturn": 0.0092,
                        "investmentCategoryIndexIPCA": 0.06
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.0718,
                        "durationDays": 120,
                        "netReturn": 0.0092,
                        "investmentCategoryIndexPostFixed": 0.1275
                    }
                ]
            };

            expect(result).toEqual(expectedSourceResult);

        });

        it('should calculate source and benchmarks correctly for LCA/LCI with pre_fixed category', () => {
            const sampleData: IFixedIncomeBenchmark = {
                investmentType: 'LCA/LCI',
                investmentCategory: 'pre_fixed',
                grossReturnPercentage: .0927,
                durationDays: 60,
                ...baseData
            };

            const result: Result = calculateBenchmark(sampleData);
            
            const expectedSourceResult = {
                "source": {
                    "investmentType": "LCA/LCI",
                    "investmentCategory": "pre_fixed",
                    "grossReturnPercentage": 0.0927,
                    "durationDays": 60,
                    "netReturn": 0.0927
                },
                "benchmark": [
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.1196,
                        "durationDays": 60,
                        "netReturn": 0.0927,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": 0.0596,
                        "durationDays": 60,
                        "netReturn": 0.0927,
                        "investmentCategoryIndexIPCA": 0.06,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.9381,
                        "durationDays": 60,
                        "netReturn": 0.0927,
                        "investmentCategoryIndexPostFixed": 0.1275,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": 0.0327,
                        "durationDays": 60,
                        "netReturn": 0.0927,
                        "investmentCategoryIndexIPCA": 0.06
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.7271,
                        "durationDays": 60,
                        "netReturn": 0.0927,
                        "investmentCategoryIndexPostFixed": 0.1275
                    }
                ]
            };

            expect(result).toEqual(expectedSourceResult);

        });

        it('should calculate source and benchmarks correctly for LCA/LCI with inflation_linked category', () => {
            const sampleData: IFixedIncomeBenchmark = {
                investmentType: 'LCA/LCI',
                investmentCategory: 'inflation_linked',
                grossReturnPercentage: .0927,
                durationDays: 60,
                ...baseData
            };

            const result: Result = calculateBenchmark(sampleData);
            
            const expectedSourceResult = {
                "source": {
                    "investmentType": "LCA/LCI",
                    "investmentCategory": "inflation_linked",
                    "grossReturnPercentage": 0.0927,
                    "durationDays": 60,
                    "netReturn": 0.1527
                },
                "benchmark": [
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.197,
                        "durationDays": 60,
                        "netReturn": 0.1527,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": 0.137,
                        "durationDays": 60,
                        "netReturn": 0.1527,
                        "investmentCategoryIndexIPCA": 0.06,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 1.5454,
                        "durationDays": 60,
                        "netReturn": 0.1527,
                        "investmentCategoryIndexPostFixed": 0.1275,
                        "tax_range": "0_180_days",
                        "tax_rate": 0.225
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.1527,
                        "durationDays": 60,
                        "netReturn": 0.1527
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 1.1976,
                        "durationDays": 60,
                        "netReturn": 0.1527,
                        "investmentCategoryIndexPostFixed": 0.1275
                    }
                ]
            };

            expect(result).toEqual(expectedSourceResult);

        });

        it('should calculate source and benchmarks correctly for LCA/LCI with post_fixed category', () => {
            const sampleData: IFixedIncomeBenchmark = {
                investmentType: 'LCA/LCI',
                investmentCategory: 'post_fixed',
                grossReturnPercentage: .0927,
                durationDays: 500,
                ...baseData
            };
            const result: Result = calculateBenchmark(sampleData);
            
            const expectedSourceResult = {
                "source": {
                    "investmentType": "LCA/LCI",
                    "investmentCategory": "post_fixed",
                    "grossReturnPercentage": 0.0927,
                    "durationDays": 500,
                    "netReturn": 0.0118
                },
                "benchmark": [
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.0143,
                        "durationDays": 500,
                        "netReturn": 0.0118,
                        "tax_range": "361_720_days",
                        "tax_rate": 0.175
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": -0.0457,
                        "durationDays": 500,
                        "netReturn": 0.0118,
                        "investmentCategoryIndexIPCA": 0.06,
                        "tax_range": "361_720_days",
                        "tax_rate": 0.175
                    },
                    {
                        "investmentType": "CDB",
                        "investmentCategory": "post_fixed",
                        "grossReturnPercentage": 0.1124,
                        "durationDays": 500,
                        "netReturn": 0.0118,
                        "investmentCategoryIndexPostFixed": 0.1275,
                        "tax_range": "361_720_days",
                        "tax_rate": 0.175
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "pre_fixed",
                        "grossReturnPercentage": 0.0118,
                        "durationDays": 500,
                        "netReturn": 0.0118
                    },
                    {
                        "investmentType": "LCA/LCI",
                        "investmentCategory": "inflation_linked",
                        "grossReturnPercentage": -0.0482,
                        "durationDays": 500,
                        "netReturn": 0.0118,
                        "investmentCategoryIndexIPCA": 0.06
                    }
                ]
            };

            expect(result).toEqual(expectedSourceResult);

        });

        const errorBaseData: IFixedIncomeBenchmark = {
            investmentType: 'LCA/LCI',
            investmentCategory: 'post_fixed',
            investmentCategoryIndexPostFixed: .1275,
            investmentCategoryIndexIPCA: .06,
            grossReturnPercentage: .0927,
            durationDays: 500,
        };

        it('should throw an error for negative grossReturnPercentage', () => {
            const sampleData: IFixedIncomeBenchmark = {
                ...errorBaseData,
                grossReturnPercentage: -.0927,
            };

            expect(() => calculateBenchmark(sampleData)).toThrowErrorMatchingSnapshot();
        });

        it('should throw an error for negative durationDays', () => {
            const testData: IFixedIncomeBenchmark = {
                ...errorBaseData,
                durationDays: -180,
            };

            expect(() => calculateBenchmark(testData)).toThrowErrorMatchingSnapshot();
        });

        it('should throw an error for negative investmentCategoryIndexPostFixed', () => {
            const testData: IFixedIncomeBenchmark = {
                ...errorBaseData,
                investmentCategoryIndexPostFixed: -1.2,
            };

            expect(() => calculateBenchmark(testData)).toThrowErrorMatchingSnapshot(); 
        });

        it('should throw an error for negative investmentCategoryIndexIPCA', () => {
            const testData: IFixedIncomeBenchmark = {
                ...errorBaseData,
                investmentCategoryIndexIPCA: -4.5,
            };

            expect(() => calculateBenchmark(testData)).toThrowErrorMatchingSnapshot();
        });

    });

});

