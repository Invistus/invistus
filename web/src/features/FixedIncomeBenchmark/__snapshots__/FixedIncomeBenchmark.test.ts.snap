// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedIncomeBenchmark calculateBenchmark should throw an error for negative durationDays 1`] = `"Duration Days cannot be negative."`;

exports[`FixedIncomeBenchmark calculateBenchmark should throw an error for negative grossReturnPercentage 1`] = `"Gross Return Percentage cannot be negative."`;

exports[`FixedIncomeBenchmark calculateBenchmark should throw an error for negative investmentCategoryIndexIPCA 1`] = `"Investment Category Index IPCA cannot be negative."`;

exports[`FixedIncomeBenchmark calculateBenchmark should throw an error for negative investmentCategoryIndexPostFixed 1`] = `"Investment Category Index Post Fixed cannot be negative."`;
